### ISSUE TITLE
bug:Vaadin 8.1.4 buffered grid editor error

### ISSUE THREAD

**Issue by querryjk** [vaadin/framework#10146]
I have a tabsheet with 2 tabs, the first is empty, the second is a grid, if I click on the grid, the editor opens correctly, but if I have the editor open and I pulse the first tab, when I return to the grid I can not open the editor by clicking, until the screen refreshes.

PD:  If the grid is setBuffered(false) and the grid has a validation error, the editor is also locked when returning

Thank a lot

Grid<Clientes> grid= new Grid<>(Clientes.class);
grid.removeAllColumns();
grid.getEditor().setEnabled(true);
grid.getEditor().setBuffered(true);

Binder<Clientes> binder = grid.getEditor().getBinder();
Binding<Clientes, String> bind = binder.forField(new TextField()).bind(Clientes::getNombre,Clientes::setNombre);
grid.addColumn(e-> e.getNombre()).setCaption("Nombre").setEditorBinding(bind).setWidth(200);

List<Clientes> lista=new ArrayList<>();
for(int i=0;i<10;i++){
Clientes cli= new Clientes();
cli.setNombre("nombre "+i);
lista.add(cli);
}

grid.setItems(lista);

VerticalLayout tab1 = new VerticalLayout();
tab1.setCaption("tab1");

VerticalLayout tab2 = new VerticalLayout();
tab2.setCaption("tab1");
tab2.addComponent(grid);

TabSheet tabSheet= new TabSheet(tab1,tab2);
tabSheet.setSizeFull();
addComponent(tabSheet);

**Comment by tsuoanttila** [comment]
Sounds like the clean up of Editor when `detach` happens for Grid is not done correctly.

**Comment by pwilkin** [comment]
A quick look shows that it's more complex - `detach` does not get called when tabs are changed.

**Comment by pwilkin** [comment]
As in, I've added: 

```
    @Override
    public void detach() {
        if (getEditor().isOpen()) {
            getEditor().cancel();
        }
        super.detach();
    }
```

to `Grid` and it doesn't help (doesn't get triggered when changing tabs). Might be more of an issue with cleanup upon *visibility* change.

**Comment by pwilkin** [comment]
`TabSheet` generally is pretty hackish in the sense that it doesn't report to the server-side component in the hidden tab that it just got hidden. This might be a bug in itself.

**Comment by Manfred-on-github** [comment]
This fix has broken our code!

At some situation we use:
```
Registration reg = someComponent.addShortcutListener(aListener);
someComponent.addDetachListener(ev -> reg.remove() );
```

After switching between tabs our shortcut listener is gone.

Please consider a different fix for the grid editor problem. 

B.t.w. I also doubt that current fix really solves the problem. What if some uplevel container (containing the grid) is temporarily made invisible (`setVisible(false)` ) and made visible again later. 

Modify the end of the code from the original poster as follows:
```
VerticalLayout tab1 = new VerticalLayout();
tab1.setCaption("tab1");
VerticalLayout tab2 = new VerticalLayout();
tab2.setCaption("tab2");
VerticalLayout tabGrid = new VerticalLayout();
tabGrid.setCaption("GridLayout");
tabGrid.addComponent(grid);
CheckBox cb = new CheckBox("toggle vis");
cb.addValueChangeListener(val -> tabGrid.setVisible(val.getValue()));
tab2.addComponent(cb);
tab2.addComponent(tabGrid);
TabSheet tabSheet= new TabSheet(tab1,tab2);
```
and then by toggling the checkbox the original problem shows up again
